src/mlpack/methods/ann/layer/multihead_attention_impl.hpp

```diff
@@ -0,0 +1,231 @@
+/**
+ * @file methods/ann/layer/multihead_attention_impl.hpp
+ * @author Mrityunjay Tripathi
+ *
+ * Implementation of the MultiheadAttention class.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_METHODS_ANN_LAYER_MULTIHEAD_ATTENTION_IMPL_HPP
+#define MLPACK_METHODS_ANN_LAYER_MULTIHEAD_ATTENTION_IMPL_HPP
+
+// In case it hasn't yet been included.
+#include "multihead_attention.hpp"
+
+#include "../visitor/backward_visitor.hpp"
+#include "../visitor/forward_visitor.hpp"
+#include "../visitor/gradient_visitor.hpp"
+#include "../init_rules/glorot_init.hpp"
+
+namespace mlpack {
+namespace ann /** Artificial Neural Network. */ {
+
+template <class AttentionPolicy, typename InputDataType,
+          typename OutputDataType>
+MultiheadAttention<AttentionPolicy, InputDataType, OutputDataType>::
+MultiheadAttention() :
+    tgtSeqLen(0),
+    srcSeqLen(0),
+    embedDim(0),
+    numHeads(0),
+    dropout(0.0),
+    deterministic(false)
+{
+  // Nothing to do here.
+}
+
+template <class AttentionPolicy, typename InputDataType,
+          typename OutputDataType>
+MultiheadAttention<AttentionPolicy, InputDataType, OutputDataType>::
+MultiheadAttention(const size_t tgtSeqLen,
+    const size_t srcSeqLen,
+    const size_t embedDim,
+    const size_t numHeads,
+    const ElemType dropout,
+    const bool deterministic) :
+    tgtSeqLen(tgtSeqLen),
+    srcSeqLen(srcSeqLen),
+    embedDim(embedDim),
+    numHeads(numHeads),
+    dropout(dropout),
+    deterministic(deterministic)
+{
+  Log::Assert(this->embedDim % numHeads == 0,
+      "'embedDim' must be divisible by 'numHeads'");
+  this->headDim = this->embedDim / numHeads;
+  if (AttentionPolicy::allInputsEqual)
+  {
+    Log::Assert(tgtSeqLen == srcSeqLen,
+        "Target sequence and Source sequence are same for Self-Attention.");
+  }
+
+  softmaxModule = new Softmax<InputDataType, OutputDataType>();
+  dropoutModule = new Dropout<InputDataType, OutputDataType>(dropout);
```

Comment:
Pointer can give you a lot of troubles. I suggest to get rid of them.

For instance suppose the second `new` throws an exception. In that case you should catch it, delete `softmaxModule` and throw the exception forward. Otherwise there would be a memory leak.

Reply:
Sure.